code/modules/power/apc.dm

```diff
@@ -757,38 +757,55 @@
 	. = ..()
 	if(.)
 		return
-	var/mob/living/carbon/human/H = user
-	var/datum/species/ethereal/eth_species = H.dna?.species
-	if(istype(H) && istype(eth_species))
-		if(H.a_intent == INTENT_HARM)
-			if(eth_species.ethereal_charge >= ETHEREAL_CHARGE_FULL - 19)
-				to_chat(H, "<span class='warning'>Your charge is full!</span>")
+
+	if(isethereal(user))
+		var/mob/living/carbon/human/H = user
+		var/datum/species/ethereal/E = H.dna.species
+		if((H.a_intent == INTENT_HARM) && (E.drain_time < world.time))
+			if(cell.charge <= (cell.maxcharge / 2)) // if charge is under 50% you shouldn't drain it
+				to_chat(H, "<span class='warning'>The APC doesn't have much power, you probably shouldn't drain any.</span>")
 				return
-			if(cell.charge < 100)
-				to_chat(H, "<span class='warning'>The cell has no charge!</span>")
+			var/obj/item/organ/stomach/ethereal/stomach = H.getorganslot(ORGAN_SLOT_STOMACH)
+			if(!istype(stomach))
+				to_chat(H, "<span class='warning'>You can't receive charge!</span>")
+				return
+			if(stomach.crystal_charge >= ETHEREAL_CHARGE_FULL)
+				to_chat(H, "<span class='warning'>Your charge is full!</span>")
 				return
+			E.drain_time = world.time + 75
 			to_chat(H, "<span class='notice'>You start channeling some power through the APC into your body.</span>")
-			if(do_after(user, 30, target = src))
-				if(cell.charge >= 100 && eth_species.ethereal_charge < ETHEREAL_CHARGE_FULL - 19)
+			if(do_after(user, 75, target = src))
+				if(cell.charge <= (cell.maxcharge / 2) || (stomach.crystal_charge >= ETHEREAL_CHARGE_FULL))
+					to_chat(H, "<span class='warning'>You can't receive charge from the APC.</span>")
+					return
+				if(istype(stomach))

```

Comment:
Just remove this istype, it is pointless since you know stomach has to be of type since you checked it beforehand, so you know it is of type.
Additionally, if the target is qdeleted the do_after will fail anyway, so there won't be an issue